sshd-common/src/main/java/org/apache/sshd/common/config/keys/OpenSshCertificate.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sshd.common.config.keys;
+
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.util.List;
+
+public final class OpenSshCertificate implements PublicKey, PrivateKey {
```

Comment:
We try to make our code as open as possible - so

* as few `final` classes as possible
* fields should be `protected` - unless there is a getter/setter
* constructors should be `public`

Reply:
If I remove the final I get:
`Error:(25) java: (design) FinalClass: Class OpenSshCertificate should be declared as final.`

Reply:
That's because the only way to build it is via the builder. If you replace the builder with getters and setters you will no longer have this problem.

Reply:
I would prefer the builder over setters, because IMO it should not be possible to change parts of the certificate later on, wdyt?

Reply:
Let me suggest sort of a compromise:

* Define `OpenSshCertificate` to be an **interface** that exposes only getters
* Define a class `OpenSshCertificateImpl` that has getters and setters and implements `OpenSshCertificate`
* Leave all the current code that checks `instanceof OpenSshCertificate`, casts to it an uses its getters as-is
* Remove the builder and use the `OpenSshCertificateImpl` to create an `OpenSshCertificate` instance when needed

This way the code maintains its "read-only" exposure of the getters, while at the same time providing more flexibility for mocks, future extensions, other implementations, etc....